core/lib/types/src/storage_writes_deduplicator.rs

```diff
@@ -113,40 +123,65 @@ impl StorageWritesDeduplicator {
                         panic!("tried removing key: {:?} before insertion", key)
                     });
                     *field_to_change -= 1;
+                    *total_size -= value.size;
+                    if is_write_initial {
+                        *total_size -= BYTES_PER_DERIVED_KEY as usize;
+                    } else {
+                        *total_size -= BYTES_PER_ENUMERATION_INDEX as usize;
+                    }
                     updates.push(UpdateItem {
                         key,
                         update_type: UpdateType::Remove(value),
                         is_write_initial,
                     });
                 }
                 (true, Some(new_value)) => {
+                    let mut value_size =
+                        compress_with_best_strategy(log.log_query.read_value, new_value).len();
```

Comment:
since the value (in the end) is changed from initial_value -> new_value, it is more correct to:

- Substract from the previous `total_size` the size of the previously modified value there
- Add to the `total_size` the new size of the value

And the `value_size` here should be counted via comparison between `initial_value` and `new_value` and not `log.log_query.read_value` and `new_value`